### DISCUSSION TITLE
v0.3.0

### DISCUSSION THREAD

**Discussion by cw-tan** [mir-group/allegro#119] [Announcements]
### Added
- Hyperparameters `tensors_mixing_mode`, `weight_individual_irreps`, `typexbasis_mode`, `tensor_track_weight_init`, `self_edge_tensor_product`
- New Bessel basis with hyperparameters `num_bessels_per_basis`, `bessel_frequency_cutoff`, `per_edge_type_cutoff`

### Changed
- Resnet normalization scheme
- [Breaking] `latent_resnet: false` enables skip connections
- [Breaking] Default for `norm_basis_mean_shift` changed from `True` -> `False`
- [Breaking] Rename `env_embed_mul` -> `num_tensor_features`
- [Breaking] Replaced `nonscalars_include_parity` with `tensor_track_allowed_irreps`

<hr /><em>This discussion was created from the release <a href='https://github.com/mir-group/allegro/releases/tag/v0.3.0'>v0.3.0</a>.</em>

**Comment by Enry99** [comment]
Is this the major code update stated in the README, or will there be another release soon?

**Reply by cw-tan** [reply]
There's more to come, this is just to get the last `allegro` version that works with the older `nequip` infrastructure out version tagged and on PyPI.

**Comment by hkqiu** [comment]
Is there any colab notebook available like before for learning？

**Reply by cw-tan** [reply]
Alongside the `allegro` [v0.4.0 update](https://github.com/mir-group/allegro/releases/tag/v0.4.0), there is a corresponding `nequip` [v0.7.0 update](https://github.com/mir-group/nequip/releases/tag/v0.7.0). There is a Colab tutorial in the `nequip` [`README.md`](https://github.com/mir-group/nequip/blob/main/README.md), where one can learn how to use the NequIP framework. Allegro is a specific model that is used in the NequIP framework, and there's a minimal example to show how one can modify the `model` portion of the config file to use Allegro models in the NequIP framework [here](https://github.com/mir-group/allegro/blob/main/configs/tutorial.yaml).

**Comment by musadu2** [comment]
Hey everybody, I have a question about num_bessels_per_basis hyper parameter!

I used run allegro 0.2.0 and now running 0.3.0. There is this new parameter that I need to defined but I want my current trainings to be consistent with my previous runs.


What should I set the num_bessels_per_basis parameter?

 I would be very glad if you can answer this question.

**Reply by cw-tan** [reply]
It refers to the number of Bessel basis functions, usually 8 is a reasonable choice.

**Reply by musadu2** [reply]
Thanks for your answer!! I have a question about the Allegro training process. are the the configurations are wrapped internally in Allegro mlip? I mean that whether Allegro automatically wraps atomic configurations internally using periodic boundary conditions (PBCs) if PBCs are set as "T T T"?



To understand this question by myself, I tested a sample Allegro model on same wrapped/unwrapped configurations. 
The test results on unwrapped configurations:
--- Final result: ---
               f_mae =  0.229882
              f_rmse =  0.337401
             H_f_mae =  0.229882
         psavg_f_mae =  0.229882
            H_f_rmse =  0.337401
        psavg_f_rmse =  0.337401
               e_mae =  2.113169
             e/N_mae =  0.022012
              e_rmse =  2.560476
            e/N_rmse =  0.026672
            
The test results on wrapped configurations:
--- Final result: ---
               f_mae =  0.229882
              f_rmse =  0.337400
             H_f_mae =  0.229882
         psavg_f_mae =  0.229882
            H_f_rmse =  0.337400
        psavg_f_rmse =  0.337400
               e_mae =  2.113181
             e/N_mae =  0.022012
              e_rmse =  2.560488
            e/N_rmse =  0.026672
            
            
from here, I see that the differences between energy and force RMSE are on the order of ~0.000001, and I think that is a negligible difference.


I would be very glad if you can answer this question.

**Reply by musadu2** [reply]
I have been also reading the nequip source code, and here is what I found:

I found the wrapping python code inside nequip
(https://erg-illinois.slack.com/archives/D05NYBYAB2R/p1750782153463409)
it automatically wraps the atoms in the code
(https://erg-illinois.slack.com/archives/D05NYBYAB2R/p1750782277970989)
https://github.com/mir-group/nequip/blob/main/nequip/data/AtomicDataDict.py
at the beginning of this code Wrapping of atoms across periodic boundaries is encoded in the edge_cell_shift tensor
https://github.com/mir-group/nequip/blob/main/nequip/data/datamodule/_base_datamodule.py
in this code:
if "collate_fn" not in dataloader_dict:
    dataloader_dict["collate_fn"] = {
        "_target_": "nequip.data.datamodule._base_datamodule._default_collate_fn_factory"
    }
When NequIP builds a PyTorch DataLoader, it uses a custom collate_fn to assemble a batch.
The default collate_fn is _default_collate_fn_factory, which returns a function that callsAtomicDataDict.batched_from_list
AtomicDataDict.batched_from_list(...) respects edge_index, edge_cell_shift, cell, and other keys needed to track periodic images.

**Reply by Linux-cpp-lisp** [reply]
Hi @musadu2 , can you please open a new GitHub discussions thread for this on the NequIP repo? Thanks!

**Comment by musadu2** [comment]
I have a question about the Allegro training process. are the the configurations are wrapped internally in Allegro mlip? I mean that whether Allegro automatically wraps atomic configurations internally using periodic boundary conditions (PBCs) if PBCs are set as "T T T"?

To understand this question by myself, I tested a sample Allegro model on same wrapped/unwrapped configurations.
The test results on unwrapped configurations:
--- Final result: ---
f_mae = 0.229882
f_rmse = 0.337401
H_f_mae = 0.229882
psavg_f_mae = 0.229882
H_f_rmse = 0.337401
psavg_f_rmse = 0.337401
e_mae = 2.113169
e/N_mae = 0.022012
e_rmse = 2.560476
e/N_rmse = 0.026672

The test results on wrapped configurations:
--- Final result: ---
f_mae = 0.229882
f_rmse = 0.337400
H_f_mae = 0.229882
psavg_f_mae = 0.229882
H_f_rmse = 0.337400
psavg_f_rmse = 0.337400
e_mae = 2.113181
e/N_mae = 0.022012
e_rmse = 2.560488
e/N_rmse = 0.026672

from here, I see that the differences between energy and force RMSE are on the order of ~0.000001, and I think that is a negligible difference.
I have been also reading the nequip source code, and here is what I found:

I found the wrapping python code inside nequip
(https://erg-illinois.slack.com/archives/D05NYBYAB2R/p1750782153463409)
it automatically wraps the atoms in the code
(https://erg-illinois.slack.com/archives/D05NYBYAB2R/p1750782277970989)
https://github.com/mir-group/nequip/blob/main/nequip/data/AtomicDataDict.py
at the beginning of this code Wrapping of atoms across periodic boundaries is encoded in the edge_cell_shift tensor
https://github.com/mir-group/nequip/blob/main/nequip/data/datamodule/_base_datamodule.py
in this code:
if "collate_fn" not in dataloader_dict:
dataloader_dict["collate_fn"] = {
"target": "nequip.data.datamodule._base_datamodule._default_collate_fn_factory"
}
When NequIP builds a PyTorch DataLoader, it uses a custom collate_fn to assemble a batch.
The default collate_fn is _default_collate_fn_factory, which returns a function that callsAtomicDataDict.batched_from_list
AtomicDataDict.batched_from_list(...) respects edge_index, edge_cell_shift, cell, and other keys needed to track periodic images.


I would be very glad if you can answer this question.

**Reply by cw-tan** [reply]
Our models account for PBCs if provided -- see https://github.com/mir-group/nequip/blob/9eaae58cd16016d43b2999e231c9d27c21f7a0e0/nequip/nn/utils.py#L68

Both Allegro and NequIP use that utility function for computing edge vectors, which accounts for periodic images through edge_cell_shifts, which will be provided if there are PBCs.

Echoing @Linux-cpp-lisp's comment -- it's good practice to start new discussions for conceptually different topics. Also, questions should be directed to discussions, not issues. Issues are meant for bug reports, feature requests, and the like.
